### PR TITLE
Binary message handling

### PR DESCRIPTION
**Pull request by ellisjoe**
This change is [Reviewable](https://reviewable.io/reviews/palantir/http-remoting/345)

### CODE REVIEW

File: jersey-servers/src/main/java/com/palantir/remoting1/servers/jersey/BinaryMessageProvider.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    // Jersey ignores this
+    @Override
+    public long getSize(byte[] bytes, Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
+        return 0;
+    }
+
+    @Override
+    public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
+        return byte[].class.isAssignableFrom(type);
```
Comment (markelliot) [inline jersey-servers/src/main/java/com/palantir/remoting1/servers/jersey/BinaryMessageProvider.java:51]: maybe just:

    type instanceof byte[]

Comment (markelliot) [reply jersey-servers/src/main/java/com/palantir/remoting1/servers/jersey/BinaryMessageProvider.java:51]: oh nvm

File: jersey-servers/src/test/java/com/palantir/remoting1/servers/jersey/BinaryMessageProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        @Path("binary-return")
+        public byte[] showWithQueryParam(String string) {
+            return string.getBytes(StandardCharsets.UTF_8);
+        }
+
+        @POST
+        @Path("binary-consume")
+        public String byteArrayToString(byte[] bytes) {
+            return new String(bytes, StandardCharsets.UTF_8);
+        }
```
Comment (markelliot) [inline jersey-servers/src/test/java/com/palantir/remoting1/servers/jersey/BinaryMessageProviderTest.java:92]: also test HeaderParam and QueryParam byte[] use? (sadly I think these may show up in the real world)

Comment (markelliot) [reply jersey-servers/src/test/java/com/palantir/remoting1/servers/jersey/BinaryMessageProviderTest.java:92]: (also, hint: you likely need ParamConverters too)

**Comment by uschi2000** [comment]
Do we want binary data in header/query/path params?

**Comment by markelliot** [comment]
unfortunately I think this isn't a question so much of want vs. actually have.

**Comment by uschi2000** [comment]
But we can control what interfaces are allowed, at least in conjure.

**Comment by uschi2000** [comment]
Review status: 0 of 3 files reviewed at latest revision, 9 unresolved discussions.

---

*[jersey-servers/src/main/java/com/palantir/remoting1/servers/jersey/BinaryMessageProvider.java, line 29 at r1](https://reviewable.io:443/reviews/palantir/http-remoting/345#-Kc5q1Xzwgcw8bwEUAfX:-Kc5q1Y-K1xIq8rP4WpL:b-5nzp2g) ([raw file](https://github.com/palantir/http-remoting/blob/b943fb9a93b76a5564cb1d978a5dd2c327167058/jersey-servers/src/main/java/com/palantir/remoting1/servers/jersey/BinaryMessageProvider.java#L29)):*
> ```Java
> import javax.ws.rs.core.MultivaluedMap;
> import javax.ws.rs.ext.Provider;
> import org.glassfish.jersey.internal.util.Base64;
> ```

would prefer java.util.Base64, I think

---

*[jersey-servers/src/main/java/com/palantir/remoting1/servers/jersey/BinaryMessageProvider.java, line 33 at r1](https://reviewable.io:443/reviews/palantir/http-remoting/345#-Kc5p2_cGHbuT15-8B7N:-Kc5p2_cGHbuT15-8B7O:bi14756) ([raw file](https://github.com/palantir/http-remoting/blob/b943fb9a93b76a5564cb1d978a5dd2c327167058/jersey-servers/src/main/java/com/palantir/remoting1/servers/jersey/BinaryMessageProvider.java#L33)):*
> ```Java
> 
> /**
>  * Reads and writes {@code byte[]}s as base64 encoded strings.
> ```

nit: base64-encoded

---

*[jersey-servers/src/main/java/com/palantir/remoting1/servers/jersey/BinaryMessageProvider.java, line 41 at r1](https://reviewable.io:443/reviews/palantir/http-remoting/345#-Kc5q4UZNbSueAdCMoYE:-Kc5q4UZNbSueAdCMoYF:b-9bw09y) ([raw file](https://github.com/palantir/http-remoting/blob/b943fb9a93b76a5564cb1d978a5dd2c327167058/jersey-servers/src/main/java/com/palantir/remoting1/servers/jersey/BinaryMessageProvider.java#L41)):*
> ```Java
>     @Override
>     public long getSize(byte[] bytes, Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
>         return 0;
> ```

jersey's ByteArrayProvider returns bytes.length, let's do the same

---

*[jersey-servers/src/main/java/com/palantir/remoting1/servers/jersey/BinaryMessageProvider.java, line 46 at r1](https://reviewable.io:443/reviews/palantir/http-remoting/345#-Kc5qDLsIx6SKQLfLKSm:-Kc5qDLsIx6SKQLfLKSn:b-4rnaj4) ([raw file](https://github.com/palantir/http-remoting/blob/b943fb9a93b76a5564cb1d978a5dd2c327167058/jersey-servers/src/main/java/com/palantir/remoting1/servers/jersey/BinaryMessageProvider.java#L46)):*
> ```Java
>     @Override
>     public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
>         return byte[].class.isAssignableFrom(type);
> ```

jersey's implementation does types == byte[].class. which one is preferred?

---

*[jersey-servers/src/test/java/com/palantir/remoting1/servers/jersey/BinaryMessageProviderTest.java, line 31 at r1](https://reviewable.io:443/reviews/palantir/http-remoting/345#-Kc5qdH_kyoHShKas6q0:-Kc5qdH_kyoHShKas6q1:bc07tp0) ([raw file](https://github.com/palantir/http-remoting/blob/b943fb9a93b76a5564cb1d978a5dd2c327167058/jersey-servers/src/test/java/com/palantir/remoting1/servers/jersey/BinaryMessageProviderTest.java#L31)):*
> ```Java
> import javax.ws.rs.core.Application;
> import javax.ws.rs.core.MediaType;
> import org.glassfish.jersey.internal.util.Base64;
> ```

java.util.Base64

---

*[jersey-servers/src/test/java/com/palantir/remoting1/servers/jersey/BinaryMessageProviderTest.java, line 38 at r1](https://reviewable.io:443/reviews/palantir/http-remoting/345#-Kc5qh21sdD071yR2RT8:-Kc5qh21sdD071yR2RT9:b9wxmk0) ([raw file](https://github.com/palantir/http-remoting/blob/b943fb9a93b76a5564cb1d978a5dd2c327167058/jersey-servers/src/test/java/com/palantir/remoting1/servers/jersey/BinaryMessageProviderTest.java#L38)):*
> ```Java
> public final class BinaryMessageProviderTest extends JerseyTest {
> 
>     private static final String STRING = "bytes";
> ```

let's add a second test case with `byte[] BINARY_BYTES = randomBytes(100)`

---

*[jersey-servers/src/test/java/com/palantir/remoting1/servers/jersey/BinaryMessageProviderTest.java, line 63 at r1](https://reviewable.io:443/reviews/palantir/http-remoting/345#-Kc5r8QeNlAZ7UDGVoak:-Kc5r8QeNlAZ7UDGVoal:b-oeh1fh) ([raw file](https://github.com/palantir/http-remoting/blob/b943fb9a93b76a5564cb1d978a5dd2c327167058/jersey-servers/src/test/java/com/palantir/remoting1/servers/jersey/BinaryMessageProviderTest.java#L63)):*
> ```Java
>         String actual = target("/binary-consume")
>                 .request()
>                 .buildPost(Entity.json(ENCODED_BYTES))
> ```

would prefer to send the base64 encoded String here, not the byte[] for the base64-encoded String. less magic.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/http-remoting/345)*

**Comment by uschi2000** [comment]
ah, let's add an e2e test too with a JaxRsClient

**Comment by ellisjoe** [comment]
re supporting binary headers and query params: do we actually even want to support `byte[]` as the body? doesn't seem particularly useful and we probably should just support streaming

**Comment by uschi2000** [comment]
Isn't that what this PR is about?

**Comment by ellisjoe** [comment]
yeah. was talking to mark and maybe this isn't useful?

**Comment by uschi2000** [comment]
I'm fine either way. So you want to ditch your conjure PR too then?

**Comment by ellisjoe** [comment]
no, I think we should just allow binary (byte[]) as a field in data objects. not as a raw input/output type

**Comment by uschi2000** [comment]
works for me.

**Pull request closed without merging**